File: src/waifu2x-processor.cpp
#include "waifu2x.h"
#include "gpu.h"

Waifu2xProcessor::Waifu2xProcessor(ncnn::VulkanDevice* device, int index) {
	if (device) {
		this->processor = Waifu2xProcessors::Gpu;
		this->index = index;
	} else {
		this->processor = Waifu2xProcessors::Cpu;
		this->index = 0;
	}

	this->device = device;
}

Waifu2xProcessor::~Waifu2xProcessor() {
	
}

Waifu2xProcessor* Waifu2xProcessor::get(Waifu2xProcessors proc, int index) {
	Waifu2xProcessor* processor = 0;
	get_s(&processor, proc, index);
	return processor;
}

int Waifu2xProcessor::get_s(Waifu2xProcessor** processor, Waifu2xProcessors proc, int index) {
	if (!processor) return -1;
	if (ncnn::get_gpu_count() < 1 || proc == Waifu2xProcessors::Cpu) *processor = new Waifu2xProcessor(0, -1);
	else {
		if (index < 0) return -2;
 	*processor = new Waifu2xProcessor(ncnn::get_gpu_device(index), index);
	}
	return 0;
}

const ncnn::GpuInfo* Waifu2xProcessor::getInfo() {
	if (device) return &device->info;
	else return 0;
}

ncnn::VulkanDevice* Waifu2xProcessor::getDevice() {
	return device;
}

File: wrapper/csharp/Waifu2xProcessor.cs
using System;

namespace libwaifu2x {
    /// <summary>
    /// processor
    /// </summary>
    public class Waifu2xProcessor : NativeWrapperObject {
        /// <summary>
        /// internal only
        /// </summary>
        public Waifu2xProcessor(IntPtr handle) : base(handle,
            Waifu2xWrapper.Waifu2xProcessor_destroy) {

        }

        /// <summary>
        /// processor type
        /// </summary>
        public Waifu2xProcessors Processor {
            get => Waifu2xWrapper.Waifu2xProcessor_get_processer(handle);
        }

        /// <summary>
        /// Returns the GPU index, if the processor is CPU, returns -1.
        /// </summary>
        public int Index {
            get => Waifu2xWrapper.Waifu2xProcessor_get_index(handle);
        }

        /// <summary>
        /// GPU only
        /// const ncnn::GpuInfo*
        /// </summary>
        /// <returns>const ncnn::GpuInfo handle</returns>
        public IntPtr GetInfo() {
            return Waifu2xWrapper.Waifu2xProcessor_get_info(handle);
        }

        /// <summary>
        /// GPU only
        /// ncnn::VulkanDevice*
        /// </summary>
        /// <returns>ncnn::VulkanDevice handle</returns>
        public IntPtr GetDevice() {
            return Waifu2xWrapper.Waifu2xProcessor_get_device(handle);
        }

        /// <summary>
        /// Gets the processor.
        /// </summary>
        /// <param name="processors">processor</param>
        /// <param name="index">gpu only If the value is -1, <paramref name="index"/> is selected as the default GPU.</param>
        /// <exception cref="NativeException"></exception>
        public static Waifu2xProcessor Get(Waifu2xProcessors processors = Waifu2xProcessors.AUTO, int index = -1) {
            Waifu2xWrapper.Waifu2xProcessor_get_s(out var ptr, processors, index < 0 ? Waifu2xEnvironment.GetDefaultGPUIndex() : index)
                .EnsureSuccessStatus();
            return new Waifu2xProcessor(ptr);
        }
    }
}
